In [313]:
# 네이버 검색 API 예제 - 블로그 검색
import os
import sys
import urllib.request
client_id = "당신의 id"
client_secret = "당신의 secret"
encText = urllib.parse.quote("맥도날드 치즈버거")
url = "https://openapi.naver.com/v1/search/shop?query=" + encText # JSON 결과
# url = "https://openapi.naver.com/v1/search/blog.xml?query=" + encText # XML 결과
request = urllib.request.Request(url)
request.add_header("X-Naver-Client-Id",client_id)
request.add_header("X-Naver-Client-Secret",client_secret)
response = urllib.request.urlopen(request)
rescode = response.getcode()
if(rescode==200):
    response_body = response.read()
    print(response_body.decode('utf-8'))
else:
    print("Error Code:" + rescode)

{
	"lastBuildDate":"Sun, 02 Oct 2022 22:36:02 +0900",
	"total":28,
	"start":1,
	"display":10,
	"items":[
		{
			"title":"<b>맥도날드 치즈버거<\/b> 세트",
			"link":"https:\/\/search.shopping.naver.com\/gate.nhn?id=22356499740",
			"image":"https:\/\/shopping-phinf.pstatic.net\/main_2235649\/22356499740.20200401122310.jpg",
			"lprice":"4510",
			"hprice":"",
			"mallName":"네이버",
			"productId":"22356499740",
			"productType":"1",
			"brand":"",
			"maker":"",
			"category1":"여가\/생활편의",
			"category2":"생활편의",
			"category3":"e쿠폰",
			"category4":"패스트푸드"
		},
		{
			"title":"<b>맥도날드<\/b> 더블 <b>치즈버거<\/b> 세트",
			"link":"https:\/\/search.shopping.naver.com\/gate.nhn?id=22356505857",
			"image":"https:\/\/shopping-phinf.pstatic.net\/main_2235650\/22356505857.20200401122531.jpg",
			"lprice":"5760",
			"hprice":"",
			"mallName":"네이버",
			"productId":"22356505857",
			"productType":"1",
			"brand":"",
			"maker":"",
			"category1":"여가\/생활편의",
			"category2":"생활편의",
			"category3":"e쿠폰",
			"category4"

## 우리의 목표는 

#### url의 1~100페이지까지 그래픽카드 쇼핑몰 정보를 긁어오고 싶다.

##### 목표를 해결하기 위해서

1. url을 만들어주는 함수가 필요합니다.

- 쇼핑물 정보를 긁어야 하니까 url에 shop이 추가되어야 합니다.
- "검색어"를 가지는 url이 필요합니다. url의 x페이지에 y개를 보여주세요.
- x페이지에 y개를 보여주세요

2. 구현
- 함수가 받아와야 하는 값은 shop, 검색어, x, y가 필요합니다
- giveme_url(shop, 검색어, x,  y) 의 함수를 만들어 봅시다.

In [314]:
def gen_search_url(api_node, search_text, start_num, disp_num):
    base = "https://openapi.naver.com/v1/search"
    node = "/" + api_node + ".json"
    param_query = "?query=" + urllib.parse.quote(search_text)
    param_target = "&start=" + str(start_num)
    param_disp = "&display=" + str(disp_num)
    
    return base + node + param_query + param_target + param_disp

In [315]:
 gen_search_url("shop", "치즈버거", 1, 3)

'https://openapi.naver.com/v1/search/shop.json?query=%EC%B9%98%EC%A6%88%EB%B2%84%EA%B1%B0&start=1&display=3'

In [316]:
def giveme_url_error(site_type, search, start_num, show_num ):
    site = "https://openapi.naver.com/v1/search/" + site_type + '?query="' +  \
            search  + '"&start=' + str(start_num)  + "&display=" + str(show_num)
    return site

In [317]:
# 결과를 보면 알겠지만 오류가 주소에 한글이 포함되기 때문에 오류가 발생한다.
giveme_url_error("shop", "치즈버거", 1, 3)

'https://openapi.naver.com/v1/search/shop?query="치즈버거"&start=1&display=3'

In [318]:
def giveme_url(site_type, search, start_num, show_num ):
    site = "https://openapi.naver.com/v1/search/" + site_type + '?query="' +  \
            urllib.parse.quote(search)  + '"&start=' + str(start_num)  + "&display=" + str(show_num)
    return site

In [319]:
 giveme_url("shop", "치즈버거", 1, 3)

'https://openapi.naver.com/v1/search/shop?query="%EC%B9%98%EC%A6%88%EB%B2%84%EA%B1%B0"&start=1&display=3'

2. url의 코드를 뽑아내고 싶습니다.
- id와 비밀번호를 입력받고
- url의 코드를 받아옵시다.

- def giveme_code(url)

In [320]:
import json

def giveme_code(url):
    request = urllib.request.Request(url)
    request.add_header("X-Naver-Client-Id", client_id)
    request.add_header("X-Naver-Client-Secret", client_secret)
    response = urllib.request.urlopen(request)
    print("페이지를 불러왔습니다.")
    return json.loads(response.read().decode("utf-8"))

In [321]:
url =  giveme_url("shop", "치즈버거", 1, 3)
code = giveme_code(url)
code

페이지를 불러왔습니다.


{'lastBuildDate': 'Sun, 02 Oct 2022 22:36:03 +0900',
 'total': 14591,
 'start': 1,
 'display': 3,
 'items': [{'title': '햄버거 수제버거 버거 만들기 <b>치즈버거</b> 밀키트 3인분',
   'link': 'https://search.shopping.naver.com/gate.nhn?id=32269766460',
   'image': 'https://shopping-phinf.pstatic.net/main_3226976/32269766460.20220511115618.jpg',
   'lprice': '18900',
   'hprice': '',
   'mallName': '네이버',
   'productId': '32269766460',
   'productType': '1',
   'brand': '',
   'maker': '',
   'category1': '식품',
   'category2': '냉동/간편조리식품',
   'category3': '햄버거',
   'category4': ''},
  {'title': '사조대림 대림선 <b>치즈버거</b> 160g',
   'link': 'https://search.shopping.naver.com/gate.nhn?id=33861982727',
   'image': 'https://shopping-phinf.pstatic.net/main_3386198/33861982727.20220804135041.jpg',
   'lprice': '1110',
   'hprice': '',
   'mallName': '네이버',
   'productId': '33861982727',
   'productType': '1',
   'brand': '',
   'maker': '',
   'category1': '식품',
   'category2': '냉동/간편조리식품',
   'category3': '햄버거',
   'cat

In [322]:
url =  giveme_url("shop", "RTX3050", 1, 3)
code = giveme_code(url)
code

페이지를 불러왔습니다.


{'lastBuildDate': 'Sun, 02 Oct 2022 22:36:03 +0900',
 'total': 25534,
 'start': 1,
 'display': 3,
 'items': [{'title': 'MSI 지포스 <b>RTX 3050</b> 게이밍 X D6 8GB 트윈프로져8',
   'link': 'https://search.shopping.naver.com/gate.nhn?id=30835551621',
   'image': 'https://shopping-phinf.pstatic.net/main_3083555/30835551621.20220207143935.jpg',
   'lprice': '409000',
   'hprice': '',
   'mallName': '네이버',
   'productId': '30835551621',
   'productType': '1',
   'brand': 'MSI',
   'maker': 'MSI',
   'category1': '디지털/가전',
   'category2': 'PC부품',
   'category3': '그래픽카드',
   'category4': 'NVIDIA계열'},
  {'title': 'MSI 지포스 <b>RTX 3050</b> 벤투스 2X OC D6 8GB',
   'link': 'https://search.shopping.naver.com/gate.nhn?id=30973400619',
   'image': 'https://shopping-phinf.pstatic.net/main_3097340/30973400619.20220217094432.jpg',
   'lprice': '379000',
   'hprice': '',
   'mallName': '네이버',
   'productId': '30973400619',
   'productType': '1',
   'brand': 'MSI',
   'maker': 'MSI',
   'category1': '디지털/가전',
   'cate

3. 데이터 전처리하여 저장하는 함수를 작성해보자.
- 제목, 가격, 판매자, 링크를 변수로 생성하여 append를 사용해서 변수에 계속해서 추가해나가면 우리가 원하는 정보를 얻을 수 있다.

In [323]:
code['items'][0]["title"]

'MSI 지포스 <b>RTX 3050</b> 게이밍 X D6 8GB 트윈프로져8'

In [324]:
code['items'][0]["lprice"]

'409000'

In [325]:
code['items'][0]["link"]

'https://search.shopping.naver.com/gate.nhn?id=30835551621'

In [326]:
code['items'][0]["mallName"]

'네이버'

In [327]:
def del_b(str):
    str = str.replace("<b>", "")
    str = str.replace("</b>", "")
    return str

In [328]:
title = del_b(code['items'][0]["title"])
title

'MSI 지포스 RTX 3050 게이밍 X D6 8GB 트윈프로져8'

In [329]:
import pandas as pd

def save_info(data):
    title = [del_b(each["title"]) for each in data["items"]]
    link = [each["link"] for each in data["items"]]
    price = [each["lprice"] for each in data["items"]]
    mall_name = [each["mallName"] for each in data["items"]]
    maker = [each["maker"] for each in data["items"]]
    brand = [each["brand"] for each in data["items"]]
    
    result = pd.DataFrame({
        "제목" : title,
        "가격" : price,
        "제조사" : maker,
        "브랜드" : brand,
        "링크" : link,
        "판매자" : mall_name
    }, columns = ["제목", "가격", "제조사", "브랜드", "링크", "판매자" ])
    return result

In [330]:
save_info(code)

제목      가격  제조사  브랜드  \
0     MSI 지포스 RTX 3050 게이밍 X D6 8GB 트윈프로져8  409000  MSI  MSI   
1        MSI 지포스 RTX 3050 벤투스 2X OC D6 8GB  379000  MSI  MSI   
2  이엠텍 지포스 RTX 3050 STORM X Dual OC D6 8GB  404550  이엠텍  이엠텍   

                                                  링크  판매자  
0  https://search.shopping.naver.com/gate.nhn?id=...  네이버  
1  https://search.shopping.naver.com/gate.nhn?id=...  네이버  
2  https://search.shopping.naver.com/gate.nhn?id=...  네이버

## 이제 네이버 API의 힘을 볼 시간이다.

- 네이버 API의 한페이지당 100개씩 표시된다. 그러면 3050데이터 1000개를 가져온다면, 10page를 불러오면 된다.

In [331]:
result_datas = []

for n in range(1, 1000, 100):
    url = giveme_url("shop", "RTX3050", n, 100 )
    RTX_result = giveme_code(url)
    result = save_info(RTX_result)
    result_datas.append(result)
    
result_datas

페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.


[                                                 제목      가격  제조사  브랜드  \
 0              MSI 지포스 RTX 3050 게이밍 X D6 8GB 트윈프로져8  409000  MSI  MSI   
 1                 MSI 지포스 RTX 3050 벤투스 2X OC D6 8GB  379000  MSI  MSI   
 2           이엠텍 지포스 RTX 3050 STORM X Dual OC D6 8GB  404550  이엠텍  이엠텍   
 3   리드텍 WinFast 지포스 RTX 3050 HURRICANE WHITE D6 8GB  348950  리드텍  리드텍   
 4        조텍 GAMING 지포스 RTX 3050 TWIN Edge OC D6 8GB  384200   조텍   조텍   
 ..                                              ...     ...  ...  ...   
 95          리드텍 WinFast 지포스 RTX 3050 CLASSIC D6 8GB  340200  리드텍  리드텍   
 96          리드텍 WinFast 지포스 RTX 3050 CLASSIC D6 8GB  342000  리드텍  리드텍   
 97          리드텍 WinFast 지포스 RTX 3050 CLASSIC D6 8GB  342000  리드텍  리드텍   
 98          리드텍 WinFast 지포스 RTX 3050 CLASSIC D6 8GB  342000  리드텍  리드텍   
 99          리드텍 WinFast 지포스 RTX 3050 CLASSIC D6 8GB  342000  리드텍  리드텍   
 
                                                    링크      판매자  
 0   https://search.shopping.naver.com/gate.

In [332]:
result_df = pd.concat(result_datas)
result_df

제목      가격  제조사    브랜드  \
0              MSI 지포스 RTX 3050 게이밍 X D6 8GB 트윈프로져8  409000  MSI    MSI   
1                 MSI 지포스 RTX 3050 벤투스 2X OC D6 8GB  379000  MSI    MSI   
2           이엠텍 지포스 RTX 3050 STORM X Dual OC D6 8GB  404550  이엠텍    이엠텍   
3   리드텍 WinFast 지포스 RTX 3050 HURRICANE WHITE D6 8GB  348950  리드텍    리드텍   
4        조텍 GAMING 지포스 RTX 3050 TWIN Edge OC D6 8GB  384200   조텍     조텍   
..                                              ...     ...  ...    ...   
95                MSI 지포스 RTX 3050 벤투스 2X OC D6 8GB  448320         MSI   
96             갤럭시 갤라즈 GALAX 지포스 RTX 3050 V2 D6 8GB  419030  갤럭시  GALAX   
97             갤럭시 갤라즈 GALAX 지포스 RTX 3050 V2 D6 8GB  436320       GALAX   
98             갤럭시 갤라즈 GALAX 지포스 RTX 3050 V2 D6 8GB  480000  갤럭시  GALAX   
99             갤럭시 갤라즈 GALAX 지포스 RTX 3050 V2 D6 8GB  480000  갤럭시  GALAX   

                                                   링크   판매자  
0   https://search.shopping.naver.com/gate.nhn?id=...   네이버  
1   https://search.shopping.naver.com/gate.nhn?id=...   네이버  
2   https://search.shopping.naver.com/gate.nhn?id=...   네이버  
3   https://search.shopping.naver.com/gate.nhn?id=...   네이버  
4   https://search.shopping.naver.com/gate.nhn?id=...   네이버  
..                                                ...   ...  
95  https://search.shopping.naver.com/gate.nhn?id=...  11번가  
96  https://search.shopping.naver.com/gate.nhn?id=...    쿠팡  
97  https://search.shopping.naver.com/gate.nhn?id=...  11번가  
98  https://search.shopping.naver.com/gate.nhn?id=...  11번가  
99  https://search.shopping.naver.com/gate.nhn?id=...  11번가  

[1000 rows x 6 columns]

In [333]:
result_df.reset_index().tail()

index                                    제목      가격  제조사    브랜드  \
995     95     MSI 지포스 RTX 3050 벤투스 2X OC D6 8GB  448320         MSI   
996     96  갤럭시 갤라즈 GALAX 지포스 RTX 3050 V2 D6 8GB  419030  갤럭시  GALAX   
997     97  갤럭시 갤라즈 GALAX 지포스 RTX 3050 V2 D6 8GB  436320       GALAX   
998     98  갤럭시 갤라즈 GALAX 지포스 RTX 3050 V2 D6 8GB  480000  갤럭시  GALAX   
999     99  갤럭시 갤라즈 GALAX 지포스 RTX 3050 V2 D6 8GB  480000  갤럭시  GALAX   

                                                    링크   판매자  
995  https://search.shopping.naver.com/gate.nhn?id=...  11번가  
996  https://search.shopping.naver.com/gate.nhn?id=...    쿠팡  
997  https://search.shopping.naver.com/gate.nhn?id=...  11번가  
998  https://search.shopping.naver.com/gate.nhn?id=...  11번가  
999  https://search.shopping.naver.com/gate.nhn?id=...  11번가

In [334]:
result_df["가격"] = result_df["가격"].astype("float")

In [335]:
result_df.to_excel("RTX3050.xlsx", encoding="utf-8")

In [336]:
result_datas2 = []

for n in range(1, 1000, 100):
    url = giveme_url("shop", "RTX3070ti", n, 100 )
    RTX_result = giveme_code(url)
    result = save_info(RTX_result)
    result_datas2.append(result)
    
result_datas2

페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.
페이지를 불러왔습니다.


[                                                   제목       가격     제조사    브랜드  \
 0               MSI 지포스 RTX 3070 Ti 게이밍 X 트리오 D6X 8GB   869000     MSI    MSI   
 1          MSI 지포스 RTX 3070 Ti 슈프림 X D6X 8GB 트라이프로져2S   899000     MSI    MSI   
 2   갤럭시코리아 GALAX 지포스 RTX 3070 Ti EX WHITE OC D6X 8...   550000     갤럭시   엔비디아   
 3       엔비디아 리더스 시스템 RTX3070TI FE 파운더리 에디션 그래픽카드 리퍼비쉬   710000  리더스시스템   엔비디아   
 4        기가바이트 지포스 RTX 3070 Ti Gaming OC D6X 8GB 제이씨현   855800   기가바이트  기가바이트   
 ..                                                ...      ...     ...    ...   
 95              COLORFUL 지포스 RTX 3070 Ti 토마호크 D6X 8GB   873000                  
 96  갤럭시 GALAX 지포스 RTX 3070 Ti EX GAMER BLACK OC D6...  3500000     갤럭시  GALAX   
 97  새로운 다채로운 RTX 3070Ti 그래픽 카드 8GB GDDR6X RTX3070 ...  1426193                  
 98  갤럭시 Galaxy GeForce RTX 3070 Ti EX WHITE OC D6X...  1036000  갤럭시코리아  GALAX   
 99     ASUS TUF Gaming 지포스 RTX 3070 Ti O8G OC D6X 8GB   896640           ASUS   
 
              

In [337]:
result_df = pd.concat(result_datas2)
result_df.reset_index().tail()
result_df.to_excel("RTX3070ti.xlsx", encoding="utf-8")